### PR TITLE
 Search: use multi-fields for Wildcard queries 

### PR DESCRIPTION
**Pull request by stsewd**
Wildcard queries are slow (on .org it returns 502, on .com since the db
is small it works just fine).

https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl.html#query-dsl-allow-expensive-queries

These type of queries can be optimized by using the Wildcard field
https://www.elastic.co/guide/en/elasticsearch/reference/current/keyword.html#wildcard-field-type.

Currently that field isn't implemented in the ES dependencies,
but is just a subclass of the Keyword field (I'll see if I can send a PR
upstream).

As we still want to make use of the SimpleString queries,
I'm using the multi-fields feature.

This change is editing the index,
so we need to rebuild the index and re-index.

To test it locally:

- `inv docker.manage 'search_index --rebuild'`.
- `inv docker.manage reindex_elasticsearch`.
- Enable the `DEFAULT_TO_FUZZY_SEARCH` feature flag on a project.

I'm not sure how to write tests for this one,
and we can't test if this is really fast in production...
But testing it locally works and gives better results for both, sections and
domains!

This is based on https://github.com/readthedocs/readthedocs.org/pull/7582 since the wildcard field was added in ES 7.x.
As we are going to re-index when deploying 7.x, I think we can deploy this change together with that (otherwise we'll need to do the re-index again).

One example where you can see better results from domains (with the wildcard search  there is a match in the title instead of matching the content)

Screenshot_2020-10-28 Read the Docs Documentation Simplified — Read the Docs 5 6 1 documentation(1)
Screenshot_2020-10-28 Read the Docs Documentation Simplified — Read the Docs 5 6 1 documentation

### CODE REVIEW

**Comment by stsewd** [comment]
This isn't blocked anymore

**Review by kadirselcuk — APPROVED** [review]

**Review by kadirselcuk — APPROVED** [review]

**Review by humitos — COMMENTED** [review]
I don't follow too much the ES code for this change, but if we are worried about the queries being slow, shouldn't we put this behind a feature flag and roll it out slowly?

**Comment by stsewd** [comment]
@humitos this feature is already in production under a feature flag, it's slow (doing a wildcard search on a no-wildcard field), that's why this PR, using a WildcardField should make it faster.

File: readthedocs/search/documents.py [hunk 1]
```diff
@@ -31,6 +38,13 @@ def update(self, *args, **kwargs):
 @project_index.document
 class ProjectDocument(RTDDocTypeMixin, Document):
 
+    """
+    Document representation of a Project.
+
+    We use multi-fields to be able to perform other kind of queries over the same field.
+    ``raw`` fields are used for Wildcard queries.
+    """
```
Comment (ericholscher) [inline readthedocs/search/documents.py:46]: 💯

**Review by ericholscher — APPROVED** [review]
This looks like a good change, but I assume requires a reindex? I'm a little worried about this, but I guess we can always just roll back to the old index if this breaks during deploy?

**Comment by stsewd** [comment]
@ericholscher just tested this locally, we don't need to re-index during deploy, we can do it later, or even just rebuilding a project would create the new fields to make use of them (some fields won't be highlighted till we re-index but is under a feature flag, so all good). I think we could just re-build some projects to test the change, and if it looks good we can re-index everything to test it more broadly with the feature flag.

**Pull request merged**